enterprise/server/soci/soci.go

```diff
@@ -0,0 +1,316 @@
+	log.Infof("soci artifacts not found, generating them for image %s", imageName)
+	pullImage(ctx, imageName)
+	return runSoci(ctx, imageName)
+}
+
+// Pulls the requested image using containerd.
+// TODO(iain): remove containerd from the mix.
+func pullImage(ctx context.Context, imageName string) error {
+	cmd := []string{"ctr", "i", "pull", imageName}
+	start := time.Now()
+	res := commandutil.Run(ctx, &repb.Command{Arguments: cmd}, "" /*=workDir*/, nil /*=statsListener*/, nil /*=stdio*/)
+	log.Infof("Pulling image %s took %s", imageName, time.Since(start))
+	return res.Error
+}
+
+// "soci create" generates two types of artifacts that we'll want to store:
+//  1. A soci index -- this is a json file that contains the ztoc of each
+//     layer of the indexed image.
+//  2. A bunch of ztocs (<= 1 per layer) -- ztoc stands for Zip Table of
+//     Contents. This is a json file that contains a map from filename to
+//     the byte offset and size where that file exists in the indexed
+//     layer. Note: only layers above a certain size are indexed, so there
+//     may be fewer ztocs than layers.
+//
+// This command writes output like this:
+// layer sha256:9c72ff7b0a48cd9e951aea7375a148e0e57f05e56efbed1c39c6bc4758492b7b -> ztoc skipped
+// layer sha256:48bc001ee0552d9afa5b9ca0dda2135a50d28604accfa9a43c7f664283c58351 -> ztoc skipped
+// layer sha256:bda554bae2c8319953b95eeeccd1a5d960e0fed35217b22001d1576c053fc20f -> ztoc sha256:0a1c0c1f64b37bbd2f344820e1afdf7fa36d36da0caad2332b5492321246709b
+// layer sha256:ed1a99fd3d74b1a636391a4f9e6dd4afcac9aa32f61d2325371192206e86d6fa -> ztoc sha256:58ef1c215e60a09aa9f347b813291b81fd0ab43884aa5884429b7f14318cd54f
+// layer sha256:647d29b55f8f9e3af8adf57e17456ecc8874f42fe74764b0cae49f90de0091e7 -> ztoc sha256:19d4ed82863392b725b953fa219f96fd01049ec47db13ecd57b0005902ead8fd
+// layer sha256:dd7a3bff4215db2b8ffd2f74ebb970596910cefdd98714fb8913e29b364dc75a -> ztoc sha256:d15d4f6b47331f124fb60c73a8209fd5e458044d597382faf6de5043f0fd14d3
+// platform {amd64 linux  [] } -> soci index sha256:b8956274ef0fda6832bc3b2dfdafd14696534c27277e7a3294dcfb8a271131be
+//
+// Get the digests of all ztocs and the soci index (there should only be
+// one index).
+func runSoci(ctx context.Context, imageName string) (*repb.Digest, []*repb.Digest, error) {
+	log.Debugf("indexing image %s", imageName)
+	cmd := []string{"soci", "create", imageName}
+	start := time.Now()
+	res := commandutil.Run(ctx, &repb.Command{Arguments: cmd}, "" /*=workDir*/, nil /*=statsListener*/, nil /*=stdio*/)
+	log.Infof("Indexing image %s took %s", imageName, time.Since(start))
+	if res.Error != nil {
+		return nil, nil, res.Error
+	}
+	return parseSociOutput(string(res.Stdout))
+}
+
+// Parses the command output described above into a digest of the soci index
+// and a list of digests of the ztocs and returns those.
+func parseSociOutput(output string) (*repb.Digest, []*repb.Digest, error) {
```

Comment:
Having to parse the soci output is a bit unfortunate. The "create command" looks [relatively simple](https://github.com/awslabs/soci-snapshotter/blob/main/cmd/soci/commands/create.go). Have you considered using the soci APIs (same ones the create command is using) to this indexing?

Reply:
Yeah good call, changed to read from the artifact db here as discussed and added a TODO to get rid of this entirely and just call the `create()` code directly.